pkg/deployment/resources/member_cleanup.go

```diff
@@ -0,0 +1,126 @@
+//
+// DISCLAIMER
+//
+// Copyright 2018 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Ewout Prangsma
+//
+
+package resources
+
+import (
+	"context"
+	"time"
+
+	driver "github.com/arangodb/go-driver"
+	api "github.com/arangodb/kube-arangodb/pkg/apis/deployment/v1alpha"
+	"github.com/arangodb/kube-arangodb/pkg/util/k8sutil"
+)
+
+const (
+	// minMemberAge is the minimum duration we expect a member to be created before we remove it because
+	// it is not part of a deployment.
+	minMemberAge = time.Minute * 10
+)
+
+// CleanupRemovedMembers removes all arangod members that are no longer part of ArangoDB deployment.
+func (r *Resources) CleanupRemovedMembers() error {
+	// Decide what to do depending on cluster mode
+	switch r.context.GetSpec().GetMode() {
+	case api.DeploymentModeCluster:
+		if err := r.cleanupRemovedClusterMembers(); err != nil {
+			return maskAny(err)
+		}
+		return nil
+	default:
+		// Other mode have no concept of cluster in which members can be removed
+		return nil
+	}
+}
+
+// cleanupRemovedClusterMembers removes all arangod members that are no longer part of the cluster.
+func (r *Resources) cleanupRemovedClusterMembers() error {
+	log := r.log
+	ctx := context.Background()
+
+	// Ask cluster for its health
+	client, err := r.context.GetDatabaseClient(ctx)
+	if err != nil {
+		return maskAny(err)
+	}
+	c, err := client.Cluster(ctx)
+	if err != nil {
+		return maskAny(err)
+	}
+	h, err := c.Health(ctx)
+	if err != nil {
+		return maskAny(err)
+	}
+
+	serverFound := func(id string) bool {
+		_, found := h.Health[driver.ServerID(id)]
+		return found
+	}
+
+	// For over all members that can be removed
+	status := r.context.GetStatus()
+	status.Members.ForeachServerGroup(func(group api.ServerGroup, list *api.MemberStatusList) error {
+		if group != api.ServerGroupCoordinators && group != api.ServerGroupDBServers {
+			// We're not interested in these other groups
+			return nil
+		}
+		for _, m := range *list {
+			if serverFound(m.ID) {
+				// Member is (still) found, skip it
+				if m.Conditions.Update(api.ConditionTypeMemberOfCluster, true, "", "") {
+					list.Update(m)
+					if err := r.context.UpdateStatus(status); err != nil {
```

Comment:
Yes, other member can be affected.
It would mean that other members may late with recording their "I'm a member of the cluster" state. If anything, they will be removed later, not earlier.